File: 0402-remove-k-digits/0402-remove-k-digits.cpp
class Solution {
public:
    string removeKdigits(string num, int k) {
        int n = num.size();
        stack<char>st;
        for(int i=0; i<n; i++) {
            while(!st.empty() && st.top() > num[i] && k > 0) {
                k--;
                st.pop();
            }
            if(st.empty() && num[i] == '0') continue;
            st.push(num[i]);
        }
        while(k > 0 && !st.empty()) {
            st.pop();
            k--;
        }
        num = "";
        while(!st.empty()) {
            num += st.top(); st.pop();
        }
        reverse(num.begin(), num.end());
        return num == "" ? "0" : num;
    }
};

File: 0402-remove-k-digits/0402-remove-k-digits.java
class Solution {
    public String removeKdigits(String num, int k) {
        Stack<Character>st = new Stack<>();
        for(int i=0; i<num.length(); i++) {
            while(!st.isEmpty() && st.peek() > num.charAt(i) && k > 0) {
                st.pop();
                k--;
            }
            if(st.isEmpty() && num.charAt(i) == '0') continue;
            st.push(num.charAt(i));
        }
        while(k > 0 && !st.isEmpty()) {
            st.pop();
            k--;
        }
        StringBuilder sb = new StringBuilder();
        while(!st.isEmpty()) {
            sb.append(st.pop());
        }
        sb.reverse();
        return sb.isEmpty() ? "0" : sb.toString();
    }
}